src/server/journal/streamer.cc

```diff
@@ -244,14 +249,21 @@ RestoreStreamer::~RestoreStreamer() {
 void RestoreStreamer::Cancel() {
   auto sver = snapshot_version_;
   snapshot_version_ = 0;  // to prevent double cancel in another fiber
+  fiber_cancelled_ = true;
   if (sver != 0) {
-    fiber_cancelled_ = true;
     db_slice_->UnregisterOnChange(sver);
     JournalStreamer::Cancel();
   }
 }
 
 bool RestoreStreamer::ShouldWrite(const journal::JournalItem& item) const {
+  if (item.cmd == "FLUSHALL" || item.cmd == "FLUSHDB") {
+    // On FLUSH* we restart the migration
+    CHECK(dest_ != nullptr);
+    dest_->Shutdown(SHUT_RDWR);
```

Comment:
I think we need another mechanism to stop migration. because we could have sent already FIN so we get data race in this case

Reply:
I think that the use case of supporting a FLUSHALL done right with a migration finalization is not really interesting. As long as it doesn't crash - we're in a good place, consistency is pretty much bonus points
having said that, what other mechanism do you have in mind?

Reply:
I think we need to provide context into restore_streamer and set error there, and set handler for error that stop migration